In [1]:
import fastText
import pandas as pd
train_set = 'dane/train.txt'
dev_set = 'dane/dev.txt'
test_set = 'dane/test.txt'

rate = 0.3

In [2]:
def count_precision(K, pred, real):
    hit = 0
    miss = 0
    for row in range(0, len(pred)):
        num = min(K, len(pred[row]))
        for element in range(0, num):
            test = (pred[row][element] in real[row])
            if test:
                hit+=1
            else:
                miss+=1
    return (hit/(hit+miss))

def count_total_precision(pred, real):
    precision = 0.0
    for K in [1,3,5]:
        precision += count_precision(K, pred, real)
    return precision/3.0

def test_model(model, test_set):
    data = pd.read_csv(dev_set, sep='jp2gmdxd', header=None)
    text = data[0].str.replace("__label__[^\n ]* ", "").str.replace(" +", " ").to_list()
    labels = data[0].str.findall("__label__[^\n ]*").to_list()
    res = model.predict(text, k=5)
    res_lists = []
    for entry in res:
        labs = []
        for row in entry:
            labs.append(row[1])
        res_lists.append(labs)
    return count_total_precision(res_lists, labels)

In [3]:
for dims in [400]:
    print("=== dim " + str(dims) + " ===")
    for ngrams in [2]:
        for epochs in [4,5,6,7,8,9,10,11,12]:
            model = fastText.train_supervised(input=train_set, epoch=epochs, lr = rate, wordNgrams = ngrams, dim = dims, minCount = 1, loss='hs')
            precision = test_model(model, dev_set)
            print("Precision on ngrams: " + str(ngrams) + ", epochs: " + str(epochs) + ": " + str(precision))
            del(model)

=== dim 400 ===


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:21: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.


Precision on ngrams: 2, epochs: 4: 0.4690228682781136
Precision on ngrams: 2, epochs: 5: 0.4719272902574829
Precision on ngrams: 2, epochs: 6: 0.473868500932573
Precision on ngrams: 2, epochs: 7: 0.4755913762491278
Precision on ngrams: 2, epochs: 8: 0.4766351736487948
Precision on ngrams: 2, epochs: 9: 0.4758707761705649
Precision on ngrams: 2, epochs: 10: 0.4771825112504465
Precision on ngrams: 2, epochs: 11: 0.4765205068179889
Precision on ngrams: 2, epochs: 12: 0.476513999100923


# new min 3
* Precision on ngrams: 2, epochs: 7: 0.47850735081536294
* Precision on ngrams: 2, epochs: 8: 0.47929805685376037
* @9 : 0.4796


# dim 400 min 2
* Precision on ngrams: 2, epochs: 5: 0.47346045952705246
* Precision on ngrams: 2, epochs: 6: 0.4752772070532709
* Precision on ngrams: 2, epochs: 7: 0.47696681138606195
* Precision on ngrams: 2, epochs: 8: 0.47794044449924833
* Precision on ngrams: 2, epochs: 9: 0.47751679742238995

# dim 400 min 3
* Precision on ngrams: 2, epochs: 5: 0.47487684588928714
* Precision on ngrams: 2, epochs: 6: 0.4762236006426908
* Precision on ngrams: 2, epochs: 7: 0.47789082905870434
* Precision on ngrams: 2, epochs: 8: 0.4785271386348046
* Precision on ngrams: 2, epochs: 9: 0.4784082961798848
* Precision on ngrams: 2, epochs: 10: 0.4781595229098928

# dim 400 min 4 - najlepszy!
* Precision on ngrams: 2, epochs: 4: 0.47417153509270155
* Precision on ngrams: 2, epochs: 5: 0.47732717215740944
* Precision on ngrams: 2, epochs: 6: 0.47905936928705634
* Precision on ngrams: 2, epochs: 7: 0.4799145838825259
* Precision on ngrams: 2, epochs: 8: 0.48089732335876495
* Precision on ngrams: 2, epochs: 9: 0.4813129955506275
* Precision on ngrams: 2, epochs: 10: 0.4808059532494095
* Precision on ngrams: 2, epochs: 11: 0.4813650091971473
* Precision on ngrams: 2, epochs: 12: 0.4815445609842636
* Precision on ngrams: 2, epochs: 13: 0.48153383024038043

# dim 400 min 5
* Precision on ngrams: 2, epochs: 5: 0.4729224851723943
* Precision on ngrams: 2, epochs: 6: 0.47484989689259266
* Precision on ngrams: 2, epochs: 7: 0.4761287644188206
* Precision on ngrams: 2, epochs: 8: 0.4759162249956665
* Precision on ngrams: 2, epochs: 9: 0.47705030013384214
* Precision on ngrams: 2, epochs: 10: 0.47636375907990663

In [2]:
model = fastText.train_supervised(input=train_set, epoch = 1, lr = 0.3, wordNgrams = 2, dim = 300, minCount = 1, loss='hs')

(('__label__rower',), array([0.0125822]))


In [22]:
for dataset in [dev_set, test_set]:
    precision = test_model(model, dataset)
    print("Precision on " + dataset + ": " + str(precision))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:21: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.


Precision on dane/dev.txt: 0.0
Precision on dane/test.txt: 0.0


In [4]:
model.save_model('candidate.bin')

* lr = 0.2 ngrams = 2 dim = 300 mincount=2 epoch 11 precision 0.4746788
* lr = 0.2 ngrams = 2 dim = 350 mincount=2 epoch 11 precision 0.4757021

* new 0.477237355 -> 0.472
* new 0.478 ->  0.474

* minn i ws nie zmieniają wyniku

* count 1 epochs 11 dim 400: 0.48054